### PR TITLE
fix(sorting): preserve values in boolean sort

### LINKED ISSUE #23: bug: Silent data corruption in sort_boolean when input contains non-{0,1} integral values
## Problem Summary
In `include/algoat/sorting/boolean_sort.hpp`, `sort_boolean` executes a single-pass zero-counting algorithm and populates the remaining slice with `1`s via `std::memset`.

If an integer array containing non-boolean values (e.g., `uint8_t` with values `{0, 2, 5, 0}`) is passed to this function, all non-zero elements are silently overwritten with `1`, resulting in catastrophic silent data corruption.

## Theoretical Invariant Violation
- **Multiset Conservation Violation:** Output multiset diverges from input multiset: $|A_{\text{out}}| = |A_{\text{in}}|$, but $\text{multiset}(A_{\text{out}}) \neq \text{multiset}(A_{\text{in}})$.

## Target Resolution
1. Constrain `sort_boolean` via C++20 concepts strictly to boolean types (`std::same_as<T, bool>`).
2. For integral types, replace unsafe `std::memset` overwriting with a branchless 2-way in-place partition (Dijkstra boolean partition) preserving the exact values.

## Target Public Interface
```cpp
namespace algoat::sorting {

template <typename T>
void sort_boolean(std::span<T> data) noexcept;

} // namespace algoat::sorting
```

## Acceptance Criteria
- [ ] Strict type constraint or two-pass verification preventing silent element overwriting.
- [ ] Unit tests added verifying conservation of multiset elements on arbitrary integer arrays.

## References & Primary Sources
- [SEI CERT C++ Coding Standard – INT04-CPP: Enforce Multi-set Invariants](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=88046682)

## Blocked By
None — can start immediately.

## 🤝 How to Claim This Issue
To claim this issue and have it automatically assigned to you:
1. Comment `/attempt` below.
2. Our automation bot will check your active issue count and assign the issue to you immediately.
3. Check out the [Contributing Guidelines](https://github.com/Sambit003/algoat/blob/main/CONTRIBUTING.md) for build, testing, and formatting standards.
4. **Tip:** Open a Draft Pull Request early to share progress, link this issue (Fixes #23), and get continuous feedback!

## Community Feedback & Alternative Proposals
We welcome feedback, design critiques, and extensions from the community. If you have suggestions for alternative algorithmic strategies, architecture-specific micro-optimizations (e.g. AVX-512 / ARM NEON / SVE), or relevant benchmark datasets, please feel free to share your thoughts in the discussion below.

### PR DESCRIPTION
**Pull request by system-99**
Fixes #23.

The previous Boolean sort implementation used `std::memset` to write `1`
values. This silently overwrote arbitrary nonzero integral values.

This change:

- Replaces `std::memset` with in-place two-way partitioning.
- Places zero values before nonzero values.
- Preserves every original element and duplicate.
- Adds regression tests for arbitrary integral values.

## Type of Change

- [x] `fix`: Bug fix
- [x] `test`: Added regression tests
- [x] `docs`: Updated function documentation

## Example:
Input:  {0, 2, 0, 5}
Output: {0, 0, 2, 5}

```text
|Aout| = |Ain|
multiset(Aout) = multiset(Ain)

## Tests

- Boolean-sort tests: passed
- Full C++ test suite: 192 tests passed

## Commands run:
cmake --build --preset debug --target test_boolean_sort
ctest --test-dir build/debug -R BooleanSortTest --output-on-failure
ctest --preset debug

### CODE REVIEW

File: include/algoat/sorting/boolean_sort.hpp [hunk 1]
```diff
@@ -9,44 +9,50 @@
...
- * Performs a single pass counting zeros, followed by two hardware-accelerated
- * @c std::memset calls. Bypasses all comparison instructions and achieves >20x speedup
- * over standard comparison sorts.
+ * Moves all zero values before nonzero values while preserving every
+ * original element. The relative order of elements is not guaranteed.
  *
+ * For example:
+ * @code
+ * {0, 2, 0, 5} -> {0, 0, 2, 5}
+ * @endcode
```
Comment (Sambit003) [inline include/algoat/sorting/boolean_sort.hpp]: The name `sort_boolean` contradicts what this function actually does. For general integral types (e.g., `{5, 0, -3, 2}` → `{0, 0, 2, 5, -3, 100, -7}`), this is a zero-partitioner, not a sort. Either rename this function to something accurate (e.g. `partition_zero `/ `boolean_partition`), or keep `sort_boolean` strictly constrained to `bool` (or `{0, 1}` types) and expose the integral partitioning under its own identifier. The issue #23 advices to contrain towards bool data type only. So, that's what you should do.

File: include/algoat/sorting/boolean_sort.hpp [hunk 2]
```diff
@@ -9,44 +9,50 @@
...
-inline void sort_boolean(std::span<uint8_t> data) noexcept {
-    if (data.empty())
-        return;
-    size_t count_false = 0;
-    for (uint8_t val : data) {
-        if (val == 0) {
-            count_false++;
+
+template <std::integral T>
+inline void sort_boolean(std::span<T> data) noexcept {
```
Comment (Sambit003) [inline include/algoat/sorting/boolean_sort.hpp]: Replacing the entire implementation with a generic two-pointer partition discards the O(N) memset fast path for genuine `bool` / `{0, 1}` data, which was the original rationale for this utility. Constrain the memset optimization with `if constexpr (std::is_same_v<T, bool>)` (or a dedicated `std::span<bool>` overload) and fall back to in-place partitioning only for wider integral types.

File: include/algoat/sorting/boolean_sort.hpp [hunk 3]
```diff
@@ -9,44 +9,50 @@
...
-        std::memset(data.data(), 0, count_false);
-    }
-    if (count_true > 0) {
-        std::memset(data.data() + count_false, 1, count_true);
+        if(left<right){
+            std::swap(data[left],data[right-1]);
+            left++;
+            right--;
+        }
     }
```
Comment (Sambit003) [inline include/algoat/sorting/boolean_sort.hpp]: This nested loop structure incurs heavy branch misprediction penalties on uniformly distributed random inputs. A branchless partition (e.g., single-pass Lomuto with conditional increment or standard branchless rotation) will perform significantly better.

File: tests/sorting/test_boolean_sort.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    algoat::sorting::sort_boolean(std::span{data});
+
+    // Same number of elements
+    EXPECT_EQ(data.size(), original.size());
+
+    // Same values, including duplicates
+    std::sort(original.begin(), original.end());
+    std::sort(data.begin(), data.end());
+
+    EXPECT_EQ(data, original);
```
Comment (Sambit003) [inline tests/sorting/test_boolean_sort.cpp]: `PreservesSizeAndMultiset` is a duplicate of `PreservesMultiset`. Both initialize `{0, 2, 0, 5, -3, 1, 5}` and sort the vector before calling `EXPECT_EQ(data, original)`. `EXPECT_EQ` on `std::vector` already validates `size()`. Drop `PreservesSizeAndMultiset` and add test cases for all-zero arrays, all-nonzero arrays, and single-element inputs instead.

File: tests/sorting/test_boolean_sort.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+}
```
Comment (Sambit003) [inline tests/sorting/test_boolean_sort.cpp]: Add a new line at the end of file.

**Review by Sambit003 — COMMENTED** [review]
@system-99 Did you even bother reading issue #23, or did you just skim the title and start randomly typing? Because this patch makes no sense at all. You took a specialized utility whose entire reason for existing was hardware-accelerated `std::memset` speed for boolean arrays, completely ripped out that fast path, and replaced it with a garbage nested `while` loop that will thrash the branch predictor on any real-world data. And what on earth is this interface? You left the function named `sort_boolean`, widened it to take any `std::integral`, and now it turns `{5, 0, -3, 2}` into `{0, 0, 2, 5, -3, ...}`. That is not sorting. That is an in-place zero partition. Calling that `sort_boolean` is a complete lie to anyone calling this API. To top it off, you didn't even run the formatter and you committed duplicate, copy-pasted tests that test the exact same thing twice. Stop writing code until you actually understand the problem you are trying to solve. Go back to the issue, read what was actually asked, ask questions if you don't understand it, and don't send patches that destroy performance and mangle basic semantics.

**Comment by system-99** [comment]
You are right—I misunderstood the scope of issue #23. I will restore the
Boolean-only memset fast path, remove the generic integral partitioning,
delete the duplicate/inapplicable tests, and run clang-format and the focused
tests again.

**Comment by Sambit003** [comment]
I'm closing this PR as it's already messed up, the commit message and PR title + descriptions are not in proper order. 
@system-99 Do setup pre-commit hook. Everything is written in the CONTRIBUTING.md file, go through that.

**Comment by Sambit003** [comment]
After careful consideration, I've decided to reopen this PR. 

@system-99 can you please fix the latest commit message and force push here with lease? don't create another commit.

**Comment by Sambit003** [comment]
@system-99 Closing this PR. You're taking a much longer time on this, and also you've messed up the worktree.

**Comment by Sambit003** [comment]
@system-99 On a serious note, if you're uncertain about some approach and also getting confused, please do ask questions in the first place, before writing / changing reqd part of the codebase.
